libraries/AC_Avoidance/AP_OADatabase.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Common/Location.h>
+
+#if !HAL_MINIMIZE_FEATURES
+#include <AP_Param/AP_Param.h>
+#include <AP_HAL/utility/RingBuffer.h>
+
+class AP_OADatabase {
+public:
+
+    AP_OADatabase();
+
+    /* Do not allow copies */
+    AP_OADatabase(const AP_OADatabase &other) = delete;
+    AP_OADatabase &operator=(const AP_OADatabase&) = delete;
+
+    // get singleton instance
+    static AP_OADatabase *get_singleton() {
+        return _singleton;
+    }
+
+    struct OA_DbItem {
+        // TODO: turn this into Vector2Int to save memory
+        Location loc;
+        uint32_t timestamp;
+    };
+
+    void init();
+    void update();
+
+    // push a location into the database
+    void queue_push(const Location &loc, const uint32_t timestamp);
+
+    bool healthy() const { return _initialised; }
+
+    // fetch an item in database. Undefined result when i >= _database.count.
+    const OA_DbItem get_item(uint32_t i) const { return _database.items[i]; }
```

Comment:
This should be const OA_DbItem& get_item(... to avoid copying.

Reply:
yes, good idea. Missed that. Fixed!